core/services/blockheaderfeeder/block_header_feeder.go

```diff
@@ -0,0 +1,255 @@
+	if err != nil {
+		return errors.Wrap(err, "finding earliest blocknumber with blockhash")
+	}
+
+	lggr.Debugw("found earliest block number with blockhash", "earliestStoredBlockNumber", earliestStoredBlockNumber)
+
+	if earliestStoredBlockNumber == nil {
+		// store earliest blockhash and return
+		// on next iteration, earliestStoredBlockNumber will be found and
+		// will make progress in storing blockhashes using blockheader.
+		// In this scenario, f.stored is not updated until the next iteration
+		// because we do not know which block number will be stored in the current iteration
+		err = f.bhs.StoreEarliest(ctx)
+		if err != nil {
+			return errors.Wrap(err, "storing earliest")
+		}
+		lggr.Info("Stored earliest block number")
+		return nil
+	}
+
+	// get the block range from (earliestStoredBlockNumber - 1) (inclusive) to minBlockNumber (inclusive) in descending order
+	blocks, err := blockhashstore.DecreasingBlockRange(earliestStoredBlockNumber.Sub(earliestStoredBlockNumber, big.NewInt(1)), minBlockNumber)
+	if err != nil {
+		return err
+	}
+
+	// use 1 sending key for all batches because ordering matters for StoreVerifyHeader
+	fromAddress, err := f.gethks.GetRoundRobinAddress(f.chainID, blockhashstore.SendingKeys(f.fromAddresses)...)
+	if err != nil {
+		return errors.Wrap(err, "getting round robin address")
+	}
+
+	for i := 0; i < len(blocks); i += int(f.storeBlockhashesBatchSize) {
+		j := i + int(f.storeBlockhashesBatchSize)
+		if j > len(blocks) {
+			j = len(blocks)
+		}
+		blockRange := blocks[i:j]
+		blockHeaders, err := f.blockHeaderProvider.RlpHeadersBatch(ctx, blockRange)
+		if err != nil {
+			return errors.Wrap(err, "fetching block headers")
+		}
+
+		lggr.Debugw("storing block headers", "blockRange", blockRange)
+		err = f.batchBHS.StoreVerifyHeader(ctx, blockRange, blockHeaders, fromAddress)
+		if err != nil {
+			return errors.Wrap(err, "store block headers")
+		}
+		for _, blockNumber := range blockRange {
+			f.stored[blockNumber.Uint64()] = struct{}{}
+		}
+	}
+
+	if f.lastRunBlock != 0 {
+		// Prune stored, anything older than fromBlock can be discarded
+		for block := f.lastRunBlock - uint64(f.lookbackBlocks); block < fromBlock; block++ {
+			if _, ok := f.stored[block]; ok {
+				delete(f.stored, block)
+				lggr.Debugw("Pruned block from stored cache",
+					"block", block)
+			}
+		}
+	}
+	// lastRunBlock is only used for pruning
+	// only time we update lastRunBlock is when the run reaches completion, indicating
+	// that new block has been stored
+	f.lastRunBlock = latestBlockNumber
+	return nil
+}
+
+func (f *BlockHeaderFeeder) findLowestBlockNumberWithoutBlockhash(ctx context.Context, lggr logger.Logger, blockToRequests map[uint64]map[string]struct{}) *big.Int {
+	var min *big.Int
+	for block, unfulfilledReqs := range blockToRequests {
+		if len(unfulfilledReqs) == 0 {
+			continue
+		}
+		if _, ok := f.stored[block]; ok {
```

Comment:
Do we prune this in-memory mapping at any point?

Reply:
yeah we prune blocks that are older than (current block - lookback block). this logic can be found at the end of `Run` method in this file.